src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -200,24 +207,72 @@ RouterExecutorRun(QueryDesc *queryDesc, ScanDirection direction, long count, Tas
 		InstrStartNode(queryDesc->totaltime);
 	}
 
-	if (operation == CMD_INSERT || operation == CMD_UPDATE ||
-		operation == CMD_DELETE)
+	estate->es_processed = 0;
+
+	/* startup the tuple receiver */
+	if (sendTuples)
 	{
-		int32 affectedRowCount = ExecuteDistributedModify(task);
-		estate->es_processed = affectedRowCount;
+		(*destination->rStartup)(destination, operation, queryDesc->tupDesc);
 	}
-	else if (operation == CMD_SELECT)
+
+	/*
+	 * If query has not yet been executed, do so now. The main reason why the
+	 * query might already have been executed is cursors.
+	 */
+	if (!routerState->eof_underlying)
+	{
+		bool resultsOK = false;
+		bool isModificationQuery = false;
+
+		if (operation == CMD_INSERT || operation == CMD_UPDATE ||
+			operation == CMD_DELETE)
+		{
+			isModificationQuery = true;
+		}
+		else if (operation != CMD_SELECT)
+		{
+			ereport(ERROR, (errmsg("unrecognized operation code: %d",
+								   (int) operation)));
```

Comment:
On 2016-06-07 05:31:56 -0700, Önder Kalacı wrote:
> > +	 * query might already have been executed is cursors.
> > +	 */
> > +	if (!routerState->eof_underlying)
> > +	{
> > +		bool resultsOK = false;
> > +		bool isModificationQuery = false;
> > +
> > +		if (operation == CMD_INSERT || operation == CMD_UPDATE ||
> > +			operation == CMD_DELETE)
> > +		{
> > +			isModificationQuery = true;
> > +		}
> > +		else if (operation != CMD_SELECT)
> > +		{
> > +			ereport(ERROR, (errmsg("unrecognized operation code: %d",
> > +								   (int) operation)));
> 
> Minor: We already have this check on `RouterExecutorRun() -> CommutativityRuleToLockMode()`. Should we keep it here as well?

I think a bit of paranoia doesn't hurt here.